### PR TITLE
Cancel Unsaved Settings Shall Restore Original Settings

### PR DESCRIPTION
**Pull request by Chopinsky**
As noted in issue #365, if canceling unsaved changes, not all settings will be restored to their previous values. This could cause problems if users accidentally changed some settings (e.g. checked a checkbox without noticing that) but don't have a way to back out what they have changed.

Now, new settings view will always check if a change has been made, and if so, when user click "Cancel" button to close the modal view, they will be asked if they want to discard the changes -- if they choose to proceed, all changes will be restored to the previous values; otherwise, the settings modal view will stay.

Be noted that `BasicAutoComplete.cs` isn't actually changed at all, but VS 2017 behaved weird again by changing all the spacing again, which made it look like being altered, but this PR does not change any of its code (but please double check and feel free to revert any unintended code changes).

### CODE REVIEW

File: src/JuliusSweetland.OptiKey/UI/ViewModels/ManagementViewModel.cs [hunk 1]
```diff
@@ -112,9 +112,29 @@ private void Ok(Window window)
             }
         }
 
-        private static void Cancel(Window window)
+        private void Cancel(Window window)
         {
-            window.Close();
+            if (ChangesRequireRestart)
```
Comment (JuliusSweetland) [inline src/JuliusSweetland.OptiKey/UI/ViewModels/ManagementViewModel.cs]: ChangesRequireRestart does not detect if ANY changes have been made, only if changes which require OptiKey to be restarted have been made. Should this not check to see if any changes exist?

**Comment by JuliusSweetland** [comment]
@Chopinsky Just leaving another comment in case you weren't alerted to my comment feedback above.

File: src/JuliusSweetland.OptiKey/UI/ViewModels/ManagementViewModel.cs (hunk 1 above)
Comment (Chopinsky) [reply src/JuliusSweetland.OptiKey/UI/ViewModels/ManagementViewModel.cs]: My mistake, I thought that property represents if any changes have been made in the settings modal view. In this case, I think we would probably just warn users regardless to keep the scope manageable.

**Comment by kmcnaught** [comment]
I just took a look at this proposed fix, and we now get the problem that if you change some settings, save them, and then later you cancel a change, you will lose all the earlier changes you made (and confirmed) as the settings file isn't saved until the app closes.

Previously the strategy was for each tab in the management console to apply it's changes only on ApplyChanges() when the management console was "OK-ed", This seems to work well so long as the sub-views respect the protocol - currently Communikate breaks this by making immediate changes to the settings rather than storing locally in the viewmodel until ApplyChanges() is called.

Is it possible to add a test that checks each view upholds this contract? Then we could just fix Communikate settings and any future breaking changes would be flagged by the test.

**Pull request merged**